### PR TITLE
Add Option.PROFILE_SPECIFIC to profile specific PropertySources from Vault

### PR DESCRIPTION
**Pull request by marbon87**
Fixes gh-1997

### CODE REVIEW

**Comment by ryanjbaxter** [comment]
There are a number of test failures can you take a look?

**Comment by marbon87** [comment]
Hi @ryanjbaxter, i fixed the formatting failures in the new test class, but i have noch idea why the others tests are failing on circleci.
If i run `maven verify` on my machine, all tests are passing.

**Comment by marbon87** [comment]
Hi @ryanjbaxter, the test is also failing on the main branch. The class org.springframework.security.web.access.WebInvocationPrivilegeEvaluator is missing. I think it is related to last commits from @spencergibb 
image

**Comment by ryanjbaxter** [comment]
Have you pulled in the latest changes?  Main is not passing on circleci.
https://app.circleci.com/pipelines/github/spring-cloud/spring-cloud-config/510/workflows/faecde66-b87f-4a2c-bb33-180cc20057ee/jobs/4120

**Comment by marbon87** [comment]
Looks good now, i rebased on main.

File: spring-cloud-config-sample/src/test/resources/logback-test.xml [hunk 1]
```diff
@@ -0,0 +1,20 @@
+<?xml version="1.0" encoding="UTF-8" ?>
```
Comment (ryanjbaxter) [inline spring-cloud-config-sample/src/test/resources/logback-test.xml:1]: Why is this needed?

File: spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java [hunk 2]
```diff
@@ -161,8 +162,9 @@ else if (ALL_OPTIONS.size() > 2) {
 							// https://github.com/spring-cloud/spring-cloud-config/issues/1874
 							for (String profile : resource.getAcceptedProfiles()) {
 								// TODO: switch to match
-								if (propertySourceName.contains("-" + profile + ".")) {
-									// TODO: switch to Options.with() when implemented
+								if (propertySourceName.contains("-" + profile + ".") || propertySourceName
+										.matches("^" + PREFIX + "vault:.*" + "," + profile + "$")) {
```
Comment (ryanjbaxter) [inline spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java:166]: Can you add a comment about why we need this?  It will help with understanding he code

File: spring-cloud-config-sample/src/test/resources/logback-test.xml (hunk 1 above)
Comment (marbon87) [reply spring-cloud-config-sample/src/test/resources/logback-test.xml:1]: Otherwise Testcontainers is logging in DEBUG mode, which leads to a lot of debug logs from docker

File: pom.xml [hunk 3]
```diff
@@ -122,6 +122,13 @@
 				<artifactId>wiremock-jre8</artifactId>
 				<version>${wiremock.version}</version>
 			</dependency>
+			<dependency>
```
Comment (ryanjbaxter) [inline pom.xml:125]: @robertmcnees can you take a look at this and make sure it aligns with the work you did with TestContainers?

File: spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java (hunk 2 above)
Comment (spencergibb) [reply spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java:166]: I'm very hesitant to add anything repository specific. We should find a more generic way

Comment (marbon87) [reply spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java:166]: The safest way would be to add an additional argument _profile_ to the property source at the server side and match that on the client against the active profiles. But this is a bigger refactoring. I don't know if you like to make such big changes in a minor or patch release. 

Another option would be to simply check if the propert source name contains the profile name. But i do not think this is a good idea, as there can be false positives.

Comment (marbon87) [reply spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java]: > 

extracted the condition to a private method and add documentation.

File: pom.xml (hunk 3 above)
Comment (robertmcnees) [reply pom.xml:125]: Is there a reason to import testcontainers twice?  The pom pulls in 1.15.1 already.  Could we change `${testcontainers.version}` instead?

File: spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java (hunk 2 above)
Comment (ryanjbaxter) [reply spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java]: Why can't we just look for the pattern `,profile_`?

Comment (marbon87) [reply spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java]: We can check for `,profile` (without `_`).
The code can look like this:

```
if(propertySourceName.contains(".*[-,]" + profile + ".*")){
...
}
```

File: pom.xml (hunk 3 above)
Comment (marbon87) [reply pom.xml:125]: Oh sorry, missed that. I will update the property.

File: spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java (hunk 2 above)
Comment (ryanjbaxter) [reply spring-cloud-config-client/src/main/java/org/springframework/cloud/config/client/ConfigServerConfigDataLoader.java]: Let's go with that as it's similar to the logic already there

**Review by ryanjbaxter — APPROVED** [review]
This looks good to me.  @spencergibb what do you think?

**Comment by ryanjbaxter** [comment]
Can you make the PR agains the 3.0.x branch?

**Comment by ryanjbaxter** [comment]
Merged in https://github.com/spring-cloud/spring-cloud-config/commit/b5ac67167135fd6301843446ccacb68194a2f8f9
